contracts/blender/src/contract.rs

```diff
@@ -0,0 +1,199 @@
+
+    type Result<T> = core::result::Result<T, BlenderError>;
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct Blender {
+        /// Merkle tree holding all the notes.
+        notes: MerkleTree<1024>,
+        /// All the seen Merkle roots (including the current).
+        merkle_roots: Set<MerkleRoot>,
+        /// Set of presented nullifiers.
+        nullifiers: Set<Nullifier>,
+
+        /// List of registered (supported) token contracts.
+        registered_tokens: Mapping<TokenId, AccountId>,
+
+        /// Mister Blendermaster (contract admin).
+        blendermaster: AccountId,
+    }
+
+    impl Default for Blender {
+        fn default() -> Self {
+            Self::new()
+        }
+    }
+
+    impl Blender {
+        /// Instantiate contract. Set caller as blendermaster.
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            ink_lang::utils::initialize_contract(|blender: &mut Self| {
+                blender.blendermaster = Self::env().caller();
+            })
+        }
+
+        /// Trigger deposit action (see ADR for detailed description).
+        #[ink(message, selector = 1)]
+        pub fn deposit(
+            &mut self,
+            token_id: TokenId,
+            value: TokenAmount,
+            note: Note,
+            proof: Vec<u8>,
+        ) -> Result<()> {
+            self.acquire_deposit(token_id, value)?;
+            self.verify_deposit(token_id, value, note, proof)?;
+            let leaf_idx = self
+                .notes
+                .add(note)
+                .map_err(|_| BlenderError::TooManyNotes)?;
+            self.merkle_roots.insert(self.notes.root().unwrap(), &());
+
+            self.env().emit_event(Deposited {
+                token_id,
+                value,
+                note,
+                leaf_idx,
+            });
+
+            Ok(())
+        }
+
+        /// Transfer `deposit` tokens of type `token_id` from the caller to this contract.
+        fn acquire_deposit(&self, token_id: TokenId, deposit: TokenAmount) -> Result<()> {
+            let token_contract = self
+                .registered_token_address(token_id)
+                .ok_or(BlenderError::TokenIdNotRegistered)?;
+
+            build_call::<super::blender::Environment>()
+                .call_type(Call::new().callee(token_contract))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(PSP22_TRANSFER_FROM_SELECTOR))
+                        .push_arg(self.env().caller())
+                        .push_arg(self.env().account_id())
+                        .push_arg(deposit as Balance)
+                        .push_arg::<Vec<u8>>(vec![]),
+                )
+                .call_flags(CallFlags::default().set_allow_reentry(true))
+                .returns::<core::result::Result<(), PSP22Error>>()
+                .fire()
+                .map_err(|e| BlenderError::InkEnv(format!("{:?}", e)))?
+                .map_err(BlenderError::Psp22)
```

Comment:
nitpick: if you implement `TryFrom`s you can replace both with `??`.

Reply:
but then I have to return `Ok(())` explicitly :P anyway, I guess that's the correct way; changed in https://github.com/Cardinal-Cryptography/aleph-node/pull/718/commits/12dd4b0db57d5c3024727067216aa9f5839e7bc4